modules/core/src/main/java/org/apache/fluo/core/oracle/OracleServer.java

```diff
@@ -322,11 +326,29 @@ public synchronized void start() throws Exception {
       Thread.sleep(200);
     }
 
-    leaderSelector = new LeaderSelector(curatorFramework, ZookeeperPath.ORACLE_SERVER, this);
+    final InetSocketAddress addr = startServer();
+
     String leaderId = HostUtil.getHostName() + ":" + addr.getPort();
-    leaderSelector.setId(leaderId);
+    leaderLatch = new LeaderLatch(curatorFramework, ZookeeperPath.ORACLE_SERVER, leaderId);
     log.info("Leader ID = " + leaderId);
-    leaderSelector.start();
+    execService = Executors.newSingleThreadExecutor(new FluoThreadFactory("Oracle Server Worker"));
+    leaderLatch.addListener(new LeaderLatchListener() {
+      @Override
+      public void notLeader() {
+        isLeader = false;
+
+        if (started) {
+          // if we stopped the server manually, we shouldn't halt
+          Halt.halt("Oracle has lost leadership unexpectedly and is now halting.");
```

Comment:
This halt is *very* unfriendly to running Fluo services in a framework. It makes sense only if it is only ever executed as it's own Java process... and I don't think it's a good idea to make that assumption, since it constrains how it should be run.

Is there an alternative to doing this which might be available?

Reply:
It does this when the Oracle loses leadership without stop being called.  Given that the Oracle is a multithreaded application and that its very hard to reason about what those threads are doing and have control over them, halting is the best option for correctness of the system.  Also, the normal way to run the Oracle is in its own process.

Reply:
If we ever want to support running the Oracle in any other execution framework, I think it's worth trying to ensure those threads are properly managed and interrupted. But, we don't need to support that today, I suppose.